src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/x64/sysv/CallArranger.java

```diff
@@ -612,19 +503,47 @@ protected BindingCalculator(boolean forArguments) {
         }
     }
 
-    private static TypeClass classifyLayout(MemoryLayout type) {
-        List<ArgumentClassImpl> classes = classifyType(type);
-        try {
-            if (type instanceof ValueLayout) {
-                return TypeClass.ofValue(classes);
-            } else if (type instanceof GroupLayout) {
-                return TypeClass.ofStruct(classes);
-            } else {
-                throw new IllegalArgumentException("Unhandled type " + type);
+    private static List<ArgumentClassImpl>[] groupByEightBytes(GroupLayout group) {
+        long offset = 0L;
+        int nEightbytes = (int)Utils.alignUp(group.byteSize(), 8) / 8;
+        @SuppressWarnings({"unchecked", "rawtypes"})
+        List<ArgumentClassImpl>[] groups = new List[nEightbytes];
+        for (MemoryLayout l : group.memberLayouts()) {
+            groupByEightBytes(l, offset, groups);
+            if (group.isStruct()) {
+                offset += l.byteSize();
             }
-        } catch (UnsupportedOperationException e) {
-            System.err.println("Failed to classify layout: " + type);
-            throw e;
+        }
+        return groups;
+    }
+
+    private static void groupByEightBytes(MemoryLayout l, long offset, List<ArgumentClassImpl>[] groups) {
+        if (l instanceof GroupLayout) {
+            GroupLayout group = (GroupLayout)l;
+            for (MemoryLayout m : group.memberLayouts()) {
+                groupByEightBytes(m, offset, groups);
+                if (group.isStruct()) {
+                    offset += m.byteSize();
+                }
+            }
```

Comment:
This loop is duplicated in the other method above. Might want to move to helper method.